lib/internal/quic/core.js

```diff
@@ -0,0 +1,3343 @@
+  const session = this[owner_symbol];
+  if (session) {
+    process.nextTick(
+      emit.bind(
+        session,
+        'pathValidation',
+        res === NGTCP2_PATH_VALIDATION_RESULT_FAILURE ? 'failure' : 'success',
+        local,
+        remote));
+  }
+}
+
+function onSessionUsePreferredAddress(address, port, family) {
+  const session = this[owner_symbol];
+  session[kUsePreferredAddress]({
+    address,
+    port,
+    type: family === AF_INET6 ? 'udp6' : 'udp4'
+  });
+}
+
+// Called by the C++ internals to emit a QLog record.
+function onSessionQlog(str) {
+  if (this.qlogBuffer === undefined) this.qlogBuffer = '';
+
+  const session = this[owner_symbol];
+
+  if (session && session.listenerCount('qlog') > 0) {
+    // Emit this chunk along with any previously buffered data.
+    str = this.qlogBuffer + str;
+    this.qlogBuffer = '';
+    if (str === '') return;
+    session.emit('qlog', str);
+  } else {
+    // Buffer the data until both the JS session object and a listener
+    // become available.
+    this.qlogBuffer += str;
+
+    if (!session || this.waitingForQlogListener) return;
+    this.waitingForQlogListener = true;
+
+    function onNewListener(ev) {
+      if (ev === 'qlog') {
+        session.removeListener('newListener', onNewListener);
+        process.nextTick(() => {
+          onSessionQlog.call(this, '');
+        });
+      }
+    }
+
+    session.on('newListener', onNewListener);
+  }
+}
+
+// Called when an error occurs in a QuicSession. When this happens,
+// the only remedy is to destroy the session.
+function onSessionError(error) {
+  if (this[owner_symbol]) {
+    this[owner_symbol].destroy(error);
+  }
+}
+
+// Called by the C++ internals when a client QuicSession receives
+// a version negotiation response from the server.
+function onSessionVersionNegotiation(
+  version,
+  requestedVersions,
+  supportedVersions) {
+  if (this[owner_symbol]) {
+    this[owner_symbol][kVersionNegotiation](
+      version,
+      requestedVersions,
+      supportedVersions);
+  }
+}
+
+// Called by the C++ internals to emit keylogging details for a
+// QuicSession.
+function onSessionKeylog(line) {
+  if (this[owner_symbol]) {
+    this[owner_symbol].emit('keylog', line);
+  }
+}
+
+// Called by the C++ internals when a new QuicStream has been created.
+function onStreamReady(streamHandle, id, push_id) {
+  const session = this[owner_symbol];
+
+  // onStreamReady should never be called if the stream is in a closing
+  // state because new streams should not have been accepted at the C++
+  // level.
+  assert(!session.closing);
+
+  // TODO(@jasnell): Get default options from session
+  const uni = id & 0b10;
+  const {
+    highWaterMark,
+    defaultEncoding,
+  } = session[kGetStreamOptions]();
+  const stream = new QuicStream({
+    writable: !uni,
+    highWaterMark,
+    defaultEncoding,
+  }, session, push_id);
+  stream[kSetHandle](streamHandle);
+  if (uni)
+    stream.end();
```

Comment:
This will cause `'finish'` to be emitted which might not be what you intend, or at least it might differ from how `net.Socket` works. I don't think you need to call `stream.end()` here given https://github.com/nodejs/node/pull/32139.